storage/redis.js

```diff
@@ -0,0 +1,343 @@
+    // }
+    super(options)
+    this.options = options
+    this.invalidation = options.invalidation || false
+  }
+
+  async init() {
+    this.log = this.options.log || nullLogger
+    this.store = this.options.client
+
+    if (!this.invalidation) {
+      return
+    }
+
+    // TODO drop
+    // await this.listen(this.options.listener)
+    // this.options.invalidation.clear
+    this.heuristicClean()
+  }
+
+  /**
+   * @param {string} key
+   * @returns {undefined|*} undefined if key not found
+   */
+  async get(key) {
+    this.log.debug({ msg: 'acd/storage/redis.get', key })
+
+    try {
+      const value = await this.store.get(key)
+      if (!value) {
+        // TODO test
+        // clear references because the key could be expired (or evicted)
+        this.clearReferences(key)
+        return undefined
+      }
+      return JSON.parse(value)
+    } catch (err) {
+      this.log.error({ msg: 'acd/storage/redis.get error', err, key })
+    }
+  }
+
+  /**
+   * set value by key
+   * @param {string} key
+   * @param {*} value
+   * @param {number} ttl - ttl in seconds; zero means key will not be stored
+   * @param {?string[]} references
+   */
+  async set(key, value, ttl, references) {
+    // TODO can keys contains * or so?
+    this.log.debug({ msg: 'acd/storage/redis.set key', key, value, ttl, references })
+
+    ttl = Number(ttl)
+    if (!ttl || ttl < 0) {
+      return
+    }
+
+    try {
+      await this.store.set(key, stringify(value), 'EX', ttl)
+
+      if (!references) {
+        return
+      }
+
+      if (!this.invalidation) {
+        this.log.warn({ msg: 'acd/storage/redis.set, invalidation is disabled, references are useless' })
+        return
+      }
+
+      const writes = []
+
+      // clear old references
+      const currentReferences = await this.store.smembers('k:' + key)
+      this.log.debug({ msg: 'acd/storage/redis.set current references', key, currentReferences })
+      if (currentReferences.length > 1) {
+        currentReferences.sort()
+        references.sort()
+        const referencesToRemove = findNotMatching(references, currentReferences)
+
+        // remove key in current references
+        for (const reference of referencesToRemove) {
+          writes.push(['srem', 'r:' + reference, key])
+        }
+        writes.push(['del', 'k:' + key])
+      }
+
+      // TODO we can probably get referencesToAdd and referencesToRemove in a single loop
+      const referencesToAdd = currentReferences.length > 1 ? findNotMatching(currentReferences, references) : references
+      this.log.debug({ msg: 'acd/storage/redis.set references to add', key, referencesToAdd })
+
+      for (let i = 0; i < referencesToAdd.length; i++) {
+        const reference = referencesToAdd[i]
+        // r: -> reference->keys
+        writes.push(['sadd', 'r:' + reference, key])
+      }
+      // k: -> key-references
+      writes.push(['sadd', 'k:' + key, references])
```

Comment:
Make both `'r:' +referfence` and `'k:'+ key` in Redis expire after a TTL, i.e. https://redis.io/commands/expire. You'll need to add the `EXPIRE` command to the pipeline for every modification you do to the set.

Reply:
I don't think it's convenient to add TTL on references, because if I do that, I lose the information needed to clear the keys, so it means I should scan all the references to check if something is dangling and remove it
Instead in this way I can use them to pick only what I need to check

Reply:
Every single time you an operation on the references you reset the timer. So it should be impossible to have the references expire before all the referenced keys are still there.